File: mobile/src/lib/phone.ts
import * as Crypto from 'expo-crypto';
import { getLocales } from 'expo-localization';
import { CountryCode, getCountryCallingCode, parsePhoneNumberFromString } from 'libphonenumber-js';

// Phone canonicalisation — the client half of contact-match. Must produce the
// EXACT same E.164 string (and therefore the same SHA-256 hash) as the backend's
// services/phone.py for a given number, or a real match silently misses.
//
// Global by design: national-format contacts ("(415) 555…", "0803 123…") are
// parsed against the DEVICE's own region, so the same book of contacts works for
// a user in the US, the UK, or Nigeria. International numbers (+…) carry their own
// country code and ignore the region on both sides. NG is only the launch-market
// fallback for when the device reports no region at all — never a hard assumption.

const FALLBACK_REGION: CountryCode = 'NG';

let cachedRegion: CountryCode | null = null;

/** The device's own country, used as the default region for national-format
 *  numbers. Falls back to the launch market when the OS reports nothing. */
export function deviceRegion(): CountryCode {
  if (cachedRegion) return cachedRegion;
  try {
    const code = getLocales()[0]?.regionCode;
    cachedRegion = (code as CountryCode | undefined) ?? FALLBACK_REGION;
  } catch {
    cachedRegion = FALLBACK_REGION;
  }
  return cachedRegion;
}

/** Canonical E.164 for a raw phone string. Mirror of backend `canonical_phone`:
 *  libphonenumber first, then a digit-strip fallback so hashing stays total for
 *  the rare number neither library can parse. Returns '' for empty input. */
export function canonicalPhone(raw: string, region?: CountryCode): string {
  const s = (raw ?? '').trim();
  if (!s) return '';

  const parsed = parsePhoneNumberFromString(s, region ?? deviceRegion());
  if (parsed) return parsed.number; // E.164, e.g. +2348031234567

  const plus = s.startsWith('+');
  const digits = s.replace(/\D/g, '');
  if (!digits) return '';
  return (plus ? '+' : '') + digits;
}

/** Compose a canonical E.164 from an **explicitly chosen** country plus the
 *  national number the user typed. Unlike `canonicalPhone` (which guesses the
 *  region from the device and so mis-parses a foreign national number — e.g. a
 *  Nigerian `0813…` read as UK `+44…` on a GB phone), the country here is
 *  user-selected, so there is no wrong-region ambiguity. This is what the phone
 *  input uses for the user's OWN number. Returns '' for empty input. */
export function composeE164(country: CountryCode, nationalNumber: string): string {
  const s = (nationalNumber ?? '').trim();
  if (!s) return '';
  // A user who types a full +international number overrides the picker.
  if (s.startsWith('+')) return canonicalPhone(s);

  const parsed = parsePhoneNumberFromString(s, country);
  if (parsed) return parsed.number;

  // Fallback: strip a single national trunk 0, then prefix the dial code.
  const digits = s.replace(/\D/g, '').replace(/^0+/, '');
  if (!digits) return '';
  return '+' + getCountryCallingCode(country) + digits;
}

/** SHA-256 hex of the canonical phone — the value the server stored for a
 *  registered user. Lowercase hex, matching Python's `hexdigest()`. Returns null
 *  for input that canonicalises to nothing. */
export async function phoneHash(raw: string, region?: CountryCode): Promise<string | null> {
  const canonical = canonicalPhone(raw, region);
  if (!canonical) return null;
  return Crypto.digestStringAsync(Crypto.CryptoDigestAlgorithm.SHA256, canonical);
}

File: backend/app/services/phone.py
"""
Phone canonicalisation + hashing for contact matching.

Contact-match (see docs/mobile-integration-plan.md §6) compares a hash of each of
the user's device contacts against a stored hash of every registered phone. For a
match, both sides MUST produce the *identical* canonical string for the same number
— a contact saved as "0803 123 4567" has to land on the same value as an account
registered as "+234 803 123 4567".

We use Google's libphonenumber on both sides: `phonenumbers` here, `libphonenumber-js`
in the mobile client (`mobile/src/lib/phone.ts`). Both are ports of the same library
and produce identical E.164 for parseable numbers. For the rare input libphonenumber
can't parse, a tiny deterministic fallback (digits, keep a leading '+') is mirrored in
the TS so hashing stays total and consistent. `DEFAULT_PHONE_REGION` is the region used
to interpret national-format numbers ("0803…"); international ("+234…") ignores it.
"""

import hashlib

import phonenumbers

from app.config import settings


def canonical_phone(raw: str, default_region: str | None = None) -> str:
    """Canonical E.164 form. Mirror of the TS implementation. '' when empty."""
    region = default_region or settings.DEFAULT_PHONE_REGION
    s = (raw or "").strip()
    if not s:
        return ""
    try:
        parsed = phonenumbers.parse(s, region)
        # format (not is_valid) on purpose — we canonicalise even imperfect numbers
        # rather than reject them; matching, not validation, is the job here.
        return phonenumbers.format_number(parsed, phonenumbers.PhoneNumberFormat.E164)
    except phonenumbers.NumberParseException:
        plus = s.startswith("+")
        digits = "".join(ch for ch in s if ch.isdigit())
        if not digits:
            return ""
        return ("+" if plus else "") + digits


def phone_hash(raw: str, default_region: str | None = None) -> str | None:
    """SHA-256 of the canonical phone. ``None`` when there is nothing to hash."""
    canonical = canonical_phone(raw, default_region)
    if not canonical:
        return None
    return hashlib.sha256(canonical.encode("utf-8")).hexdigest()
